src/resolve/resolved-dns-packet.c

```diff
@@ -2570,6 +2571,81 @@ bool dns_packet_equal(const DnsPacket *a, const DnsPacket *b) {
         return dns_packet_compare_func(a, b) == 0;
 }
 
+int dns_packet_ede_rcode(DnsPacket *p, char **ret_ede_msg) {
+        assert(p);
+
+        _cleanup_free_ char *msg = NULL, *msg_escaped = NULL;
+        int ede_rcode = _DNS_EDNS_OPT_MAX_DEFINED;
+        int r;
+        const uint8_t *d;
+        size_t l;
+
+        if (!p->opt)
+                return -EINVAL;
+
+        d = p->opt->opt.data;
+        l = p->opt->opt.data_size;
+
+        while (l > 0) {
+                uint16_t code, length;
+
+                if (l < 4U)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "EDNS0 variable part has invalid size.");
+
+                code = unaligned_read_be16(d);
+                length = unaligned_read_be16(d + 2);
+
+                if (l < 4U + length)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "Truncated option in EDNS0 variable part.");
+
+                if (code == DNS_EDNS_OPT_EXT_ERROR) {
+                        if (length < 2U)
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                "EDNS0 truncated EDE info code.");
+                        ede_rcode = unaligned_read_be16(d + 4);
+                        r = make_cstring((char *)d + 6, length - 2U, MAKE_CSTRING_ALLOW_TRAILING_NUL, &msg);
+                        if (r < 0)
+                                return log_debug_errno(r, "Invalid EDE text in opt");
+                        else if (!utf8_is_valid(msg))
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG), "Invalid EDE text in opt");
+                        else if (ede_rcode < _DNS_EDNS_OPT_MAX_DEFINED) {
+                                msg_escaped = cescape(msg);
+                                if (!msg_escaped)
+                                        return -ENOMEM;
+                        }
+                        break;
+                }
+
+                d += 4U + length;
+                l -= 4U + length;
+        }
+
+        if (ede_rcode >= _DNS_EDNS_OPT_MAX_DEFINED)
+                return -EINVAL;
```

Comment:
why not do this "capping" here? just returning the error code as is should be safe, no? you handle unknown EDE errors in the _to_string() gracefully, hence I see no reason to eat them up here.

In particular, as more EDE errors will be defined in future. Even if we might not recognize them then, we might still want to show the EDE message though, and hence we shouldn't consider unknown EDE errors as errors, but just as valid values

Reply:
So, I think my thought when writing this was that, having seen the existing codes, the EDE aren't exactly "error" codes and more like supplementary data. The rfc also takes care to indicate that they may accompany any rcode including NOERROR, and some current EDE codes are vague or uninteresting. So I reasoned that any new codes could theoretically confer just about anything in meaning, and using their presence to select some interesting behavior is likely to be an error.

When considering how to handle the codes in resolved-dns-transaction.c though, I guess I came to the opposite conclusion, since for the purpose of feature detection I figure just about any code is evidence enough that the server understood the query properly. Looking at it now, it's clear you are right, especially since the server's error message will (hopefully) be useful to the admin. I will remove this bit.